### ISSUE TITLE
Plot: Add as an available vector, the row # of the query

### ISSUE THREAD

**Issue by struck89** [sqlitebrowser/sqlitebrowser#820]
### Details for the issue

I usualy plot x vs y, but what happens when you only want to plot y? Let me put my example:

In this image, I want to see, visually, how the variable "Total Volume" is distributed, Using only what is available:
![series](https://cloud.githubusercontent.com/assets/18152095/19331568/690a583a-90b3-11e6-826d-14bdd1ba67ac.PNG)
But it would be very useful to also have this:
![series2](https://cloud.githubusercontent.com/assets/18152095/19331716/ea788626-90b3-11e6-9347-dc7f80b4f0fe.PNG)

Which would be solved if this was added:
![result](https://cloud.githubusercontent.com/assets/18152095/19331799/affaf0be-90b4-11e6-939d-dcf81f2fa581.png)

### Useful extra information

Most probably there is a SQL command that generates such a column, I idon't know. I still think it would be a neat feature.

#### I'm opening this issue because:
- [ ] DB4S is crashing
- [ ] DB4S has a bug
- [X] DB4S needs a feature
- [ ] DB4S has another problem
#### I'm using DB4S on:
- [X] Windows: ( _version:_ ___ )
- [ ] Linux: ( _distro:_ ___ )
- [ ] Mac OS: ( _version:_ ___ )
- [ ] Other: ___
#### I'm using DB4S version:
- [X] 3.9.1
- [ ] 3.9.0
- [ ] Other: ___
#### I have also:
- [ ] Tried out the nightly version: https://github.com/sqlitebrowser/sqlitebrowser#windows
- [X] Searched for an existing similar issue: https://github.com/sqlitebrowser/sqlitebrowser/issues?utf8=%E2%9C%93&q=is%3Aissue%20

**Comment by justinclift** [comment]
Hmmm, this seems like it might be do-able using straight SQL.  Probably something like using a temporary table along with an auto-increment key.

I don't have the time/inclination to try it out personally just yet, but in theory (!) it should work. :smile:

**Comment by MKleusberg** [comment]
There are at least two ways to realise this in SQL:
1) Create a temporary table like with the results of your select statement. Then plot the data from that table instead of directly plotting the select results.
```sql
CREATE TEMPORARY TABLE tmp AS SELECT total_vol, jname FROM jhem, jdata WHERE jhem.jid=jdata.jid ORDER BY total_vol;
SELECT rowid,* FROM tmp; -- <- Plot this!
```

2) Modify the query to include a row counter. But this makes the query a lot more complicated.
```sql
SELECT (SELECT COUNT(*) FROM jhem x, jdata y WHERE x.jid>a.jid AND x.jid=y.jid ORDER BY y.total_vol) AS rownum, b.total_vol, a.jname FROM jhem a, jdata b WHERE a.jid=b.jid ORDER BY b.total_vol;  -- <- Haven't tested this but you might get the idea.
```

Still, I think @struck89's suggestion to have it in the UI makes some sense. None of these approaches is very userfriendly and I am almost certain that I made a mistake or two in the SQL for the second approach.

**Comment by struck89** [comment]
@MKleusberg the first one works like charm (still, not user friendly, on excel you can select just a column of data and ask for plotting series).

The second one adds a row which is probably the original order of the data. The order by clause has to be somewhere else, I guess inside the parenthesis.
![not the same numbers](https://cloud.githubusercontent.com/assets/18152095/19351507/1a84d7b8-912a-11e6-9f80-4fe9e163d546.PNG)
Since the row number and the rownum column are not the same, the result is not exactly what I was looking for.

**Comment by struck89** [comment]
Holly cow,
That is already KIND of implemented:
![image](https://cloud.githubusercontent.com/assets/18152095/19521840/3498adce-95e3-11e6-97da-8006f8c659f9.png)
Ok, this is strange. I read what that thing does and it is doing something unrelated. (bug)
The interesting thing is that it retrieved a set of columns (that DO NOT appear on the results tab) from what looks like a previous query.
The query I was working on has nothing to do with that set of rows and columns, but clicking on that it added the _rowid_
If I go back to sql, ask for anything else, and click that gear sign again, it happens again, retrieves that magical data from IDontKnowWhere, with my beloved _rowid_ column

**Comment by MKleusberg** [comment]
@struck89 Sorry for the delay but I believe this is fixed now :smiley:

The gear button you discovered was actually buggy. It is supposed to load all data (for large datasets only a part is loaded, so there would be data missing in the plot) but it did this only for the Browse Data tab. When there was a plot based on a manual SQL query like you had it would load the data from the Browse Data tab and show that. This data included the rowid but was not connected to your query in any way. This is fixed now: the gear button should always load the correct data now.

I've also added a 'Row #' column to the column selection view. In simple cases this is the same as the rowid but is way more reliable for what you want :smiley:

Do you have the time to give tomorrow's nightly build a quick test and report back if it's working for you? :smiley:

**Comment by struck89** [comment]
@MKleusberg Hey! no need to ask for pardon!! ;)

So, if you tell me the nightly you mention is here:
https://nightlies.sqlitebrowser.org/win64/
Yes, I'll give it a try.

Sadly though, I am using the DB Browser since I started around August to focus more on organizing the data I was generating for a project and analyzing it. This has been a temporary position and I will be unemployed at the end of November. I am sure I will be around sqlite db browser (I have read so many names, that I don't really know which is the last acronym you use 😆 )  in the near future...

I am sure this will be a great addition for more people to come!!

**Comment by struck89** [comment]
@MKleusberg 
Since you didn't confirm that, I completely forgot to check again. Today I needed it, and tried on the nightlies how it works.
It is just what I asked, I think this thread can be closed!

I hope other people find it useful as well!!
Thanks, thanks!!

**Comment by MKleusberg** [comment]
@struck89 Sorry to hear that you're unemployed now :crying_cat_face: But thank you so much for opening all those issues concerning the plot feature :+1:  We'll try our best to implement those features one by one in the near-ish future.

Also thank you for testing the nightly build and confirming that it's working now :smiley: I'll close this issue then.

**Comment by MKleusberg** [comment]
By the way, out of curiosity: what sort of data were you working on there? I was wondering, especially when seeing the loop in the 'order by' issue :smiley:

**Comment by struck89** [comment]
Hey @MKleusberg !
I have to say that I found other issues with that nighlty build. For example that between queries that had as an output the same columns, it did not remember the ones that were being plotted.
Before I had column1 as X and column2 as Y and after running a new query the graph got updated. With the nightly I had to select the columns to be plotted again.
Also I remember (it's been a month without playing with it), a significant decrease in performance when selecting the columns to plot. I don't know if that was a consequence of the feature I wished it had, or by some other addition..

And those where PV loops. Pressure-Volume loops of the left ventricle of a simulation of a human heart!
90% of what I was plotting were Pressure vs Time or Pressure vs Volume of all different heart chambers; the most iconic being the left ventricle.

Best regards and merry xmas!
